lib/promoted/ruby/client/options.rb

```diff
@@ -0,0 +1,215 @@
+        end
+
+        def user_id
+          return @user_id if @user_id
+          @user_id = request.dig(:user_info, :user_id)
+          @user_id ||= request.dig('user_info', 'user_id')
+          @user_id
+        end
+
+        def session_id
+          @session_id
+        end
+
+        #A list of the response Insertions.  This list should be truncated
+        #based on limit.
+        def insertion
+          @insertion
+        end
+
+        def log_user_id
+          return @log_user_id if @log_user_id
+          @log_user_id   = request.dig(:user_info, :log_user_id)
+          @log_user_id ||= request.dig('user_info', 'log_user_id')
+          @log_user_id
+        end
+
+        #A way to turn off logging.  Defaults to true.
+        def enabled?
+          @enabled
+        end
+
+         #Performs extra dev checks.  Safer but slower.  Defaults to true.
+        def perform_checks?
+          @perform_checks
+        end
+
+
+        # Default values to use on DeliveryRequests.
+        def default_request_values
+          @default_request_values
+        end
+
+        # Required as a dependency so clients can load reduce dependency on multiple
+        # uuid libraries.
+        def uuid
+          @uuid
+        end
+
+        # Defaults to 250ms
+        def delivery_timeout_millis
+          @delivery_timeout_millis
+        end
+
+        # Defaults to 3000ms
+        def metrics_timeout_millis
+          @metrics_timeout_millis
+        end
+
+        # For testing.  Allows for easy mocking of the clock.
+        def now_millis
+          @now_millis
+        end
+
+        def only_Log
+          @only_Log
+        end
+
+        def full_insertion
+          @full_insertion
+        end
+
+        def user_info
+          {
+            user_id: user_id,
+            log_user_id: log_user_id
+          }
+        end
+
+        def timing
+          @timing = {
+            client_log_timestamp: client_log_timestamp,
+            event_api_timestamp: event_api_timestamp
+          }
+        end
+
+        def request_id
+          @request_id
+        end
+
+        def log_request_params
+          {
+            platform_id: platform_id,
+            user_info: user_info,
+            timing: timing,
+            request: request_params(include_insertion: false),
+            insertion: compact_insertions
+          }
+        end
+
+        def request_params include_insertion: true
+          # return @request_params if @request_params
+          request
+          @request_params = {
+            platform_id: platform_id,
+            user_info: user_info,
+            timing: timing,
+            request_id: request_id,
+            view_id: view_id,
+            session_id: session_id,
+            insertion: compact_insertions
+          }
+          @request_params.merge!({insertion: compact_insertions}) if include_insertion
+          @request_params
+        end
+
+
+        def compact_insertions
```

Comment:
In the Typescript client, there are two compact functions that the client can configure that strips out fields when sending to promoted.

It looks like it's possible to pass in a function as an argument in Ruby.

I think we'll have some standard compact function (e.g. strip out everything but properties).  We'll also want to support some custom cases.

Reply:
Yeah, we can do that.